### PR TITLE
feat: configure sub command supports multiple profiles

### PR DESCRIPTION
**Pull request by nic-6443**
Signed-off-by: Nic <qianyong@api7.ai>

### CODE REVIEW

**Review by mscb402 — APPROVED** [review]

**Review by Hazel6869 — APPROVED** [review]

File: cmd/configure/cmd.go [hunk 1]
```diff
@@ -88,9 +83,49 @@ func NewCommand() *cobra.Command {
...
+			if err != nil {
+				output.Verbosef("there is no configuration file, create a new one")
+				configuration = &persistence.CloudConfiguration{}
+				// generate a random profile name if not specified at first time
+				if profileName == "" {
+					profileName = namesgenerator.GetRandomName(0)
+				}
+			} else {
+				if profileName == "" {
+					profileName = configuration.DefaultProfile
```
Comment (tokers) [inline cmd/configure/cmd.go:98]: What does that mean? Why use a default name when configure a new profile?

File: cmd/configure/cmd.go [hunk 2]
```diff
@@ -88,9 +83,49 @@ func NewCommand() *cobra.Command {
...
+				output.Errorf(err.Error())
+			}
+
 			output.Infof("successfully configured api7 cloud access token, your account is %s", me.Email)
 		},
 	}
 
+	cmd.PersistentFlags().StringVar(&options.Global.Configure.Addr, "addr", "https://api.api7.cloud", "Specify the API7 Cloud server address")
+	cmd.PersistentFlags().StringVar(&options.Global.Configure.Profile, "profile", "", "Specify the profile name")
+	cmd.PersistentFlags().BoolVar(&options.Global.Configure.Default, "set-default", true, "Set the profile as default")
```
Comment (tokers) [inline cmd/configure/cmd.go:127]: True?

**Review by tokers — COMMENTED** [review]
Missing docs for profile.

File: cmd/configure/cmd.go (hunk 1 above)
Comment (nic-6443) [reply cmd/configure/cmd.go:98]: For updating default profile.

File: cmd/configure/cmd.go (hunk 2 above)
Comment (nic-6443) [reply cmd/configure/cmd.go:127]: Yes, new profile set as default profile by default.

File: cmd/configure/cmd.go [hunk 3]
```diff
@@ -88,9 +83,49 @@ func NewCommand() *cobra.Command {
 				output.Errorf("failed to request api7 cloud: %s", err)
 			}
 
+			profileName := options.Global.Configure.Profile
+
+			configuration, err := persistence.LoadConfiguration()
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply cmd/configure/cmd.go:89]: Fixed.

File: cmd/configure/cmd.go (hunk 1 above)
Comment (nic-6443) [reply cmd/configure/cmd.go:103]: Add a comment for this.

File: cmd/configure/cmd.go (hunk 2 above)
Comment (tokers) [reply cmd/configure/cmd.go:132]: I think now let's keep the default value to false.

File: docs/configuring-cloud-cli.md [hunk 4]
```diff
@@ -64,6 +64,26 @@ WARNING: your access token will expire at 2024-01-31T00:03:50+08:00
...
+> Note Cloud CLI saves the access token to `$HOME/.api7cloud/config`.
 
-Now you can run other protected commands. Enjoy your journey!
+Configure Multiple Profiles for Cloud CLI
+----------------------------------------
+
+`cloud-cli configure` uses `https://api.api7.cloud` as the server address for
+API7 Cloud by default. If your control plane is created on another Cloud server,
+then you can specify the server address with `-addr` and specify a name for the
+current configuration to distinguish between multiple environments with `-profile`.
```
Comment (tokers) [inline docs/configuring-cloud-cli.md:75]:
```suggestion
`cloud-cli configure` uses `https://api.api7.cloud` as the server address for
API7 Cloud by default. If your control plane is created on another region,
then you can specify the server address with `--addr` and specify a name for the
current configuration to distinguish among multiple environments with `--profile`.
```

File: internal/testutils/utils.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/api7/cloud-cli/internal/persistence"
+)
+
+// PrepareFakeConfiguration prepares a fake configuration for testing.
+func PrepareFakeConfiguration(t *testing.T) {
```
Comment (tokers) [inline internal/testutils/utils.go:26]: Write it in a test file?

File: internal/persistence/configuration.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create file in %s for api7 cloud configuration: %s", configDir, err)
+	}
+
+	write, err := file.Write(data)
+	if err != nil {
+		return fmt.Errorf("failed to write configuration to %s, %s", configDir, err)
+	}
+	if write != len(data) {
+		return fmt.Errorf("failed to write configuration to %s", configDir)
```
Comment (tokers) [inline internal/persistence/configuration.go:64]:
```suggestion
		return fmt.Errorf("failed to write configuration to %s, truncated write", configDir)
```

File: internal/persistence/configuration.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	if profileName == "" {
+		profileName = configuration.DefaultProfile
+	}
+
+	profile, err := configuration.GetProfile(profileName)
+	if err != nil {
+		return fmt.Errorf("Failed to get %s profile, Please check your configuration file: %s", profileName, configDir)
+	}
+
+	output.Verbosef("Loaded access token: %s", profile.User.AccessToken)
```
Comment (tokers) [inline internal/persistence/configuration.go:108]: Remove this line.

File: internal/persistence/configuration_test.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	profile, err := config.GetDefaultProfile()
+	assert.NoError(t, err, "get default profile")
+	assert.Equal(t, "prod-token", profile.User.AccessToken, "access token should be prod-token")
+
+	profile, err = config.GetProfile("dev")
+	assert.NoError(t, err, "get dev profile")
+	assert.Equal(t, "dev-token", profile.User.AccessToken, "access token should be dev-token")
+
+	id = uuid.NewString()
+	err = os.MkdirAll(fmt.Sprintf("/tmp/%s", id), fs.ModePerm)
```
Comment (tokers) [inline internal/persistence/configuration_test.go:72]: Do not hard code `/tmp`. Use `os.TempDir`.

File: internal/testutils/utils.go (hunk 5 above)
Comment (nic-6443) [reply internal/testutils/utils.go:27]: Functions in test file can't be exported, I add a comment for this package.

File: internal/persistence/configuration.go (hunk 7 above)
Comment (nic-6443) [reply internal/persistence/configuration.go]: Removed.

File: internal/persistence/configuration_test.go (hunk 8 above)
Comment (nic-6443) [reply internal/persistence/configuration_test.go]: Fixed.

**Review by tokers — APPROVED** [review]

**Pull request merged**